Add support for custom root page content

mitchell-then opened the pull request:
- rewrite publish routines
- support multiple directives in dox header
- support source file as root page

mitchell-then commented:
This pull request contains quite a few changes, and a medium sized rewrite of how publishing is done. I considered implementing some type of page dependency management, but ultimately decided against it as the only dependent pages right now are between the root and child pages.

Since this is a large change of how publishing works, I'm looking for feedback on how I'm trying to do things. I'm not completely convinced that my implementation of the dox header is in good shape either.

mitchell-then commented:
One problem yet to be solved: remove requirement of having a title defined in the dox config file when using a source as a root page.

jesselang commented:
Hey Mitch. I like the idea of being able to specify some content for the root page. But specifying it in the page seems like it could be prone to problems. The source options introduced so far have affected only that source (i.e. `ignore` excludes the source from being published, `omit-notice` omits a generation notice only that source, etc.), but the root page really affects all sources, as all sources become children of that root page. Also, there can logically only be one root page, which also brings up a case where a user has added the directive to multiple pages in the repo, which should not be allowed.

One possible direction to take could be to use convention over configuration. By specifying a conventional source location used by dox, no configuration would be necessary, because the source is either present or isn't, and only one source can be at that location. How do you think that would fit into your use cases?

jesselang commented on README.md:
````diff
@@ -33,15 +33,51 @@ Each markdown file will be modified with a dox header, and `.dox.yaml` will be
+
+```
+<!-- dox: <directive>,1234567890 -->
+```
````
I really like this direction. Should we refer to these as "Source Directives", and refer to "files" as "source files" in this section? Aesthetically, I'd like to see the source ID at the beginning of the directive list. We might also consider that some future source directives might be some value other than a boolean switch. But keeping with the problem at hand, could we could require that the first directive be either `ignore` or the page ID, optionally followed by any other directives?

mitchell-then replied:
- [x] `source directives` and `source files` in code
- [x] source ID or ignore at beginning of list

jesselang commented on pkg/publish.go:
```diff
@@ -104,65 +163,42 @@ func PrePublish(file string, rootID string, dryRun bool) (id string, err error)
-
+func updateContent(wiki *confluence.Wiki, src source.Source, repoRoot string, dryRun bool) (id string, err error) {
 	if src.Ignore() {
+		// TODO: what to return when a page is ignored?
```
Should be an impossible condition, yeah? So return an error as a sanity check?

mitchell-then replied:
- [x] return error since we should not be here

jesselang commented on pkg/source/source.go:
```diff
@@ -4,24 +4,30 @@ import (
+	"ignore": "ignore",
+	"omitNotice": "omit-notice",
+	"rootPage": "root-page",
+}
```
We might consider enumerating directives using constants if the number of directives continues to grow, which lets us reason about directives as integers, or even do bitwise operations. Check out https://splice.com/blog/iota-elegant-constants-golang/ if you're interested in what that could look like.

mitchell-then replied:
- [x] use constants for enumerating directives

mitchell-then commented:
> 
> 
> Hey Mitch. I like the idea of being able to specify some content for the root page. But specifying it in the page seems like it could be prone to problems. The source options introduced so far have affected only that source (i.e. `ignore` excludes the source from being published, `omit-notice` omits a generation notice only that source, etc.), but the root page really affects all sources, as all sources become children of that root page. Also, there can logically only be one root page, which also brings up a case where a user has added the directive to multiple pages in the repo, which should not be allowed.
> 
> One possible direction to take could be to use convention over configuration. By specifying a conventional source location used by dox, no configuration would be necessary, because the source is either present or isn't, and only one source can be at that location. How do you think that would fit into your use cases?

I agree, if we look for a file at a predetermined location and use that as the root page, it would solve some problems.

I think we can safely rule out using the `README.md` file at the root of the repo for this purpose as it would be present in many repos (custom root page for all repos?), and I don't think it would make sense to use the contents of a README as a root page in all cases.

So that would mean creating our own convention, such as `dox_root.md` at the root of the repo. I believe this would work well in many cases. A possible problem is that I may want all my dox source files to be in a single location, like `docs/` (or maybe `dox/`) in the repo. But I think that would be a non-issue as the root page is somewhat of an exception.

I'll go ahead and start implementing this idea. What are your thoughts on the name of the file? `dox_root.md`, `DOX_ROOT.md`, `doxRoot.md`?

jesselang commented:
> I think we can safely rule out using the `README.md` file at the root of the repo for this purpose as it would be present in many repos (custom root page for all repos?), and I don't think it would make sense to use the contents of a README as a root page in all cases.

Agreed.

> So that would mean creating our own convention, such as `dox_root.md` at the root of the repo. I believe this would work well in many cases. A possible problem is that I may want all my dox source files to be in a single location, like `docs/` (or maybe `dox/`) in the repo. But I think that would be a non-issue as the root page is somewhat of an exception.

That's certainly a realistic use case. Could `dox` check the root of the repo first, then fall back to a file matching the convention anywhere in the repo?

> I'll go ahead and start implementing this idea. What are your thoughts on the name of the file? `dox_root.md`, `DOX_ROOT.md`, `doxRoot.md`?

I'm wondering if we could simply go with `ROOT.md`, which seems descriptive, but not specific to `dox`, and maybe that's OK. I haven't seen a lot of `ROOT.md` files in repos.

mitchell-then commented on pkg/source/markdown.go:
```diff
@@ -202,3 +221,33 @@ func (m *markdown) parse(filename string, opts Opts) (err error) {
 }
+
+func (m *markdown) parseDirectives() error {
+	firstDirective, optionalDirectives := m.directives[0], m.directives[1:]
```
This approach does not work when dox is run on a "new" repo. For example, a dox header of `dox: omit-notice` is valid the first time, but this does not account for that.

mitchell-then commented on pkg/source/source.go:
```diff
@@ -4,24 +4,32 @@ import (
+const doxHeaderRegexp = `dox: (.*)`
+
+// source directives (SD)
+const (
```
I've been considering moving this idea to use a type struct instead. Instead of keeping track of which directives are defined in the file and which directives are specified separately (`m.directives` and `m.omitNotice`), then having to string them back together (`strings.Join(m.directives, ", ")`). A struct may be cleaner. We could have methods to return a header representation of the directives and to get the values of directives individually. Do you think this might be a good way of going about it?

jesselang replied:
Yeah, I think so. Were you thinking we should have a `Header` struct with functions to serialize and deserialize?

jesselang replied:
After more thought, the dox header is providing identification and directives how the the source should be published. Give me another day to think through how to best represent this.

mitchell-then replied:
Sure. I have been playing around with this for a bit but haven't come up with anything good yet. There are some difficulties, like the page ID, which I don't think should be called a directive. Then, as you have brought up, we should have a solution that could handle directives with values other than bool. 

My initial attempt used a header struct, an omitNotice struct, and an ignore struct. Header struct had fields for omitNotice and ignore. Methods would have to handle the directive structs explicitly, so this solution isn't scalable. I'm not sure how an interface would work either, since we could be dealing with different value types (`directive.Value()` -> bool, `directive.Value()` -> string)

jesselang commented on pkg/publish.go:
```diff
@@ -59,15 +60,78 @@ func PrePublish(file string, rootID string, dryRun bool) (id string, err error)
-		return
+		return err
+	}
+
```
Seems like we should explicitly create a ROOT source here, before looping over all the files.

mitchell-then replied:
This is essentially what happens. First, find all eligible source files. Out of those source files, see if one is the root page. If one exists, use it, or use the default generated page. Stub the root for the ID, then publish all files.

There is some cheating going on, since we (harmlessly) stub the root page twice. I think I prefer this way, however, as if a page dependency management system is implemented, I think it would behave in a similar way. Find all files, then publish them in an appropriate order (root page likely first). The root page being published first is hard coded here instead.

mitchell-then committed:
Add support for custom root page content

- rewrite publish routines
- support multiple directives in dox header
- support source file as root page

mitchell-then commented:
Squashed commits, and have tested with some use cases. This pull request should be in a good state to merge.
